### PR TITLE
backtester: run manager

### PR DESCRIPTION
**Pull request by gloriousCode**
image

# PR Description
Since the merge of #988, it has made #1023 incompatible with the GoCryptoTrader Backtester. This PR is the prerequisite to #1023  and adds a new run manager to allow for backtesting/livestrategy runs to be managed via GRPC. Runs are now assigned IDs and store some basic stats about their run. That way, a user via GRPC can monitor many strategies running simultaneously

I've added the following new commands:
- listAllRuns,
- startRun
- stopRun
- stopAllRuns
- clearRun
- clearAllRuns

Now you can run multiple live strategies simultaneously and choose via GRPC when to stop. This will eventually lead the way to informing live strategies of decisions it could make/changing strategy values mid-run.

I added the `clearRun` commands to be able to remove inactive backtesting runs from memory, as I could foresee memory issues if there was not a way to clear them. If a strategy is running, you are not allowed to clear the run

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## How has this been tested
- [x] go test ./... -race
- [x] golangci-lint run
- [x] coverage of all new functionality
- [x] tested that single-run still works for both offline and live based strategies

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by shazbert — CHANGES_REQUESTED** [review]
This is lovely:

```
go run ./... listallruns
{
 "runs": [
  {
   "id": "2f1bf992-6d9e-493d-82e4-d2d1ddeb2801",
   "strategy_name": "ftx-cash-carry",
   "date_loaded": "2022-09-21 13:15:54 AEST",
   "date_started": "2022-09-21 13:16:36 AEST",
   "date_ended": "2022-09-21 13:16:36 AEST",
   "closed": true
  },
  {
   "id": "04e2d649-8e8d-4add-8ccb-3f0384c13ebb",
   "strategy_name": "ftx-cash-carry",
   "date_loaded": "2022-09-21 13:16:53 AEST"
  },
  {
   "id": "1eb651d4-bffc-4e42-b3bc-5a6304c9d2c5",
   "strategy_name": "ftx-cash-carry",
   "date_loaded": "2022-09-21 13:17:00 AEST"
  },
  {
   "id": "e910e2ac-7f50-40c3-8066-e5c6dbda8c1d",
   "strategy_name": "ftx-cash-carry",
   "date_loaded": "2022-09-21 13:17:05 AEST"
  }
 ]
}
```

**Review by shazbert — APPROVED** [review]
tACK!

**Review by thrasher- — APPROVED** [review]
image
tACK, all commands! Nice stuff 👍 

As discussed, one pedantic nit I thought about was that "task(s)", "job(s) or "backtest(s)" makes more sense gramatically than "runs", but those semantics can be addressed in the next or future backtester PR.

**Review by thrasher- — APPROVED** [review]
ACK changes, nice readability improvement!

File: common/timeperiods/timeperiods.go [hunk 1]
```diff
@@ -111,7 +111,7 @@ func (t *TimePeriodCalculator) calculatePeriods() {
 		return
 	}
 	iterateDateMate := t.start
-	for !iterateDateMate.Equal(t.end) && !iterateDateMate.After(t.end) {
+	for !iterateDateMate.Equal(t.end) && iterateDateMate.Before(t.end) {
```
Comment (shazbert) [inline common/timeperiods/timeperiods.go:114]: What does mate refer to?

**Review by shazbert — CHANGES_REQUESTED** [review]
changes look good! but for some reason half runs the strategy when loaded then stops.

```
go run ./... executestrategyfromfile --dnr --path="thrasher-corp\gocryptotrader\backtester\config\strategyexamples\ftx-cash-carry.strat"
```


```
[INFO] | GRPC |  28/09/2022 16:43:23  | Ready to receive commands
[INFO] | SETUP |  28/09/2022 16:44:44  | Loading config...
[DEBUG] | ORDER |  28/09/2022 16:44:44  | Order manager starting...
[DEBUG] | ORDER |  28/09/2022 16:44:44  | Order manager started.
[DEBUG] | EXCHANGE |  28/09/2022 16:44:44  | FTX forced update of available [SPOT] pairs.
[DEBUG] | EXCHANGE |  28/09/2022 16:44:44  | FTX forced update of available [FUTURES] pairs.
[DEBUG] | EXCHANGE |  28/09/2022 16:44:44  | FTX forced update of available [FUTURES] pairs.
[DEBUG] | EXCHANGE |  28/09/2022 16:44:44  | FTX forced update of available [SPOT] pairs.
[INFO] | SETUP |  28/09/2022 16:44:44  | Setting exchange settings...
[WARN] | SETUP |  28/09/2022 16:44:44  | No credentials set for FTX, this is theoretical only
[INFO] | SETUP |  28/09/2022 16:44:44  | Loading data for FTX futures BTC-20210924...
[WARN] | SETUP |  28/09/2022 16:44:44  | No credentials set for FTX, this is theoretical only
[INFO] | SETUP |  28/09/2022 16:44:44  | Loading data for FTX spot BTC/USD...
[INFO] | CONFIG |  28/09/2022 16:44:45  | ------------------Backtester Settings------------------------
[INFO] | CONFIG |  28/09/2022 16:44:45  | ------------------Strategy Settings--------------------------
[INFO] | CONFIG |  28/09/2022 16:44:45  | Strategy: ftx-cash-carry
[INFO] | CONFIG |  28/09/2022 16:44:45  | Custom strategy variables: unset
[INFO] | CONFIG |  28/09/2022 16:44:45  | Simultaneous Signal Processing: true
[INFO] | CONFIG |  28/09/2022 16:44:45  | USD value tracking: true
[INFO] | CONFIG |  28/09/2022 16:44:45  | ------------------Funding Settings---------------------------
[INFO] | CONFIG |  28/09/2022 16:44:45  | Use Exchange Level Funding: true
[INFO] | CONFIG |  28/09/2022 16:44:45  | Initial funds for ftx spot USD: 100000
[INFO] | CONFIG |  28/09/2022 16:44:45  | ------------------futures BTC-20210924 Currency Set
[INFO] | CONFIG |  28/09/2022 16:44:45  | Exchange: ftx
[INFO] | CONFIG |  28/09/2022 16:44:45  | Taker fee: 0.0007
[INFO] | CONFIG |  28/09/2022 16:44:45  | Maker fee: 0.0002
[INFO] | CONFIG |  28/09/2022 16:44:45  | Minimum slippage percent: 100
[INFO] | CONFIG |  28/09/2022 16:44:45  | Maximum slippage percent: 100
[INFO] | CONFIG |  28/09/2022 16:44:45  | Buy rules: {MinimumSize:0 MaximumSize:0 MaximumTotal:0}
[INFO] | CONFIG |  28/09/2022 16:44:45  | Sell rules: {MinimumSize:0 MaximumSize:0 MaximumTotal:0}
[INFO] | CONFIG |  28/09/2022 16:44:45  | Can use exchange defined order execution limits: false
[INFO] | CONFIG |  28/09/2022 16:44:45  | ------------------spot BTC-USD Currency Settings---
[INFO] | CONFIG |  28/09/2022 16:44:45  | Exchange: ftx
[INFO] | CONFIG |  28/09/2022 16:44:45  | Taker fee: 0.0007
[INFO] | CONFIG |  28/09/2022 16:44:45  | Maker fee: 0.0002
[INFO] | CONFIG |  28/09/2022 16:44:45  | Minimum slippage percent: 100
[INFO] | CONFIG |  28/09/2022 16:44:45  | Maximum slippage percent: 100
[INFO] | CONFIG |  28/09/2022 16:44:45  | Buy rules: {MinimumSize:0 MaximumSize:0 MaximumTotal:0}
[INFO] | CONFIG |  28/09/2022 16:44:45  | Sell rules: {MinimumSize:0 MaximumSize:0 MaximumTotal:0}
[INFO] | CONFIG |  28/09/2022 16:44:45  | Can use exchange defined order execution limits: false
[INFO] | CONFIG |  28/09/2022 16:44:45  | ------------------Portfolio Settings-------------------------
[INFO] | CONFIG |  28/09/2022 16:44:45  | Buy rules: {MinimumSize:0 MaximumSize:0 MaximumTotal:0}
[INFO] | CONFIG |  28/09/2022 16:44:45  | Sell rules: {MinimumSize:0 MaximumSize:0 MaximumTotal:0}
[INFO] | CONFIG |  28/09/2022 16:44:45  | Leverage rules: {CanUseLeverage:true MaximumOrdersWithLeverageRatio:0 MaximumOrderLeverageRate:0 MaximumCollateralLeverageRate:0}
[INFO] | CONFIG |  28/09/2022 16:44:45  | ------------------API Settings-------------------------------
[INFO] | CONFIG |  28/09/2022 16:44:45  | Data type: candle
[INFO] | CONFIG |  28/09/2022 16:44:45  | Interval: 24h0m0s
[INFO] | CONFIG |  28/09/2022 16:44:45  | Start date: 2021-01-14 00:00:00
[INFO] | CONFIG |  28/09/2022 16:44:45  | End date: 2021-09-24 00:00:00
```

Then continues when:

```
go run ./... startallruns
{
 "runs_started": [
  "e8fd9490-5ff6-47c1-bfce-6608607323e9"
 ]
}
```

🤷

File: common/timeperiods/timeperiods.go (hunk 1 above)
Comment (gloriousCode) [reply common/timeperiods/timeperiods.go:114]: :shrug: I just ran the linter and it decided to fix these for me :smile: Though I can see from git that I wrote that :smile: I can change that if you want?

**Review by shazbert — APPROVED** [review]
Comments addressed, tACK.

**Pull request merged**